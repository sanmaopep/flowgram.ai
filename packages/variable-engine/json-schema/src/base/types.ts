/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { type BaseTypeManager } from './base-type-manager';

/**
 * Base information for TypeRegistry
 */
export interface BaseTypeRegistry {
  /**
   * type reference
   */
  type: string;
  /**
   * The inherited type. If there is an inheritance, the definition of this type will use the inherited type definition,
   * and the new definition will override the inherited definition.
   */
  extend?: string;

  /**
   *
   */
  onRegister?: () => void;
}

/**
 * TypeRegistryCreator
 */
export type TypeRegistryCreator<
  Schema,
  Registry extends BaseTypeRegistry,
  Manager extends BaseTypeManager<Schema, Registry, Manager>
> = (ctx: { typeManager: Manager }) => Partial<Registry>;
