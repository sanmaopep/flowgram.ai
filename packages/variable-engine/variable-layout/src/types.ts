import { Scope } from '@flowgram.ai/variable-core';
import { FlowNodeEntity } from '@flowgram.ai/document';

export enum FlowNodeScopeTypeEnum {
  public = 'public',
  private = 'private',
}

export interface FlowNodeScopeMeta {
  node?: FlowNodeEntity;
  type?: FlowNodeScopeTypeEnum;
}

export interface ScopeVirtualNode {
  id: string;
  flowNodeType: 'virtualNode';
}

export type ScopeChainNode = FlowNodeEntity | ScopeVirtualNode;

// 节点内部的作用域
export interface FlowNodeScope extends Scope<FlowNodeScopeMeta> {}
