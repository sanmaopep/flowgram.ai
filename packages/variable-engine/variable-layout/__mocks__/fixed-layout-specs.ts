import { FlowDocumentJSON } from '@flowgram.ai/document';

export const fixLayout1: FlowDocumentJSON = {
  nodes: [
    {
      type: 'start',
      meta: {
        isStart: true,
      },
      id: 'start',
    },
    {
      type: 'getRecords',
      id: 'getRecords_07e97c55832',
    },
    {
      type: 'loop',
      id: 'forEach_260a8f85ff2',
      blocks: [
        {
          type: 'createRecord',
          id: 'createRecord_8f85ff2c11d',
        },
        {
          type: 'dynamicSplit',
          id: 'exclusiveSplit_ff2c11d0fb4',
          blocks: [
            {
              id: 'branch_f2c11d0fb42',
            },
            {
              id: 'branch_2c11d0fb42c',
            },
          ],
        },
      ],
    },
    {
      type: 'dynamicSplit',
      id: 'exclusiveSplit_88dbf2c60ae',
      meta: {
        defaultExpanded: true,
      },
      blocks: [
        {
          id: 'branch_8dbf2c60aee',
          blocks: [
            {
              type: 'dynamicSplit',
              id: 'exclusiveSplit_a59afaadc9a',
              blocks: [
                {
                  id: 'branch_59afaadc9ac',
                },
                {
                  id: 'branch_9afaadc9acd',
                  blocks: [
                    {
                      type: 'deleteRecords',
                      id: 'deleteRecords_c32807e97c5',
                    },
                  ]
                },
              ],
            },
          ]
        },
        {
          id: 'branch_dbf2c60aee4',
          blocks: [
            {
              type: 'updateRecords',
              id: 'updateRecords_7ed2a172c32',
            },
          ]
        },
      ],
    },
    {
      type: 'end',
      id: 'end',
    },
  ],
};
