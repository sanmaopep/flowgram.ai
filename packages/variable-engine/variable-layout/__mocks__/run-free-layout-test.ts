import { describe, expect, test } from 'vitest';
import { VariableEngine } from '@flowgram.ai/variable-core';
import { ASTKind } from '@flowgram.ai/variable-core';
import {
  FlowNodeVariableData
} from '../src';
import { TestConfig, getContainer } from './container';
import { WorkflowDocument, WorkflowJSON } from '@flowgram.ai/free-layout-core';

export const runFreeLayoutTest = (testName: string, spec: WorkflowJSON, config?: TestConfig) => {
  describe(testName, async () => {
    const container = getContainer('free', config);
    const flowDocument = container.get(WorkflowDocument);
    const variableEngine = container.get(VariableEngine);

    variableEngine.onScopeChange(action => {
      const { type, scope } = action;

      // 作用域默认创建一个默认变量
      if (type === 'add') {
        scope.ast.set('/', {
          kind: ASTKind.VariableDeclaration,
          type: ASTKind.String,
          key: String(scope.id),
        });
      }
    });

    // 创建一个全局作用域
    variableEngine.createScope('testScope');

    const traverseVariableDatas = () =>
      flowDocument
        .getAllNodes()
        // TODO 包含 $ 的节点不注册 variableData
        .filter(_node => !_node.id.startsWith('$'))
        .map(_node => _node.getData(FlowNodeVariableData))
        .filter(Boolean);

    // 初始化所有节点的私有作用域
    const initAllNodePrivate = () => {
      traverseVariableDatas().forEach(_data => {
        _data.initPrivate();
      });
    };

    // 初始化所有节点的可用变量
    const printAllNodeAvailableMapping = (_scopeType: 'public' | 'private' = 'public') =>
      traverseVariableDatas().reduce((acm, _data) => {
        const scope = _data[_scopeType]!;
        acm.set(String(scope.id), scope.available.variableKeys);

        return acm;
      }, new Map<string, string[]>());

    // 初始化所有节点的覆盖作用域
    const printAllCovers = (_scopeType: 'public' | 'private' = 'public') =>
      traverseVariableDatas().reduce((acm, _data) => {
        const scope = _data[_scopeType]!;
        acm.set(
          String(scope.id),
          scope.coverScopes.map(_scope => String(_scope.id)),
        );

        return acm;
      }, new Map<string, string[]>());

    await flowDocument.fromJSON(spec);

    test('test get Deps', () => {
      expect(printAllNodeAvailableMapping()).toMatchSnapshot();
    });

    test('test get Covers', () => {
      expect(printAllCovers()).toMatchSnapshot();
    });

    test('test get Deps After Init Private', () => {
      initAllNodePrivate();
      expect(printAllNodeAvailableMapping()).toMatchSnapshot();
    });

    test('test get private scope Deps', () => {
      expect(printAllNodeAvailableMapping('private')).toMatchSnapshot();
    });

    test('test get Covers After Init Private', () => {
      expect(printAllCovers()).toMatchSnapshot();
    });

    test('test get private scope Covers', () => {
      expect(printAllCovers('private')).toMatchSnapshot();
    });

    test('test sort', () => {
      expect(variableEngine.getAllScopes({ sort: true }).map(_scope => _scope.id)).toMatchSnapshot();
    });

    config?.runExtraTest?.(container);
  });


}
