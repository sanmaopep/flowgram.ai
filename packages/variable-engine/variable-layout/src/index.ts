export { FlowNodeVariableData } from './flow-node-variable-data';
export { FreeLayoutScopeChain } from './chains/free-layout-scope-chain';
export { VariableLayoutConfig } from './variable-layout-config';
export { FixedLayoutScopeChain } from './chains/fixed-layout-scope-chain';
export {
  type FlowNodeScopeMeta,
  type FlowNodeScope,
  FlowNodeScopeTypeEnum as FlowNodeScopeType,
} from './types';
export { GlobalScope, bindGlobalScope } from './scopes/global-scope';
