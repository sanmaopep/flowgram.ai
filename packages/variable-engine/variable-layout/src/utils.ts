/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { FlowNodeEntity } from '@flowgram.ai/document';

import { FlowNodeVariableData } from './flow-node-variable-data';

/**
 * Use `node.scope` instead
 * @deprecated
 * @param node
 */
export function getNodeScope(node: FlowNodeEntity) {
  return node.getData(FlowNodeVariableData).public;
}

/**
 * Use `node.privateScope` instead
 * @deprecated
 * @param node
 */
export function getNodePrivateScope(node: FlowNodeEntity) {
  return node.getData(FlowNodeVariableData).initPrivate();
}
