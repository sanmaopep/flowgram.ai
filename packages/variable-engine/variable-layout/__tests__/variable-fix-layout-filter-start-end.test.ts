import { FlowNodeScope } from '../src/types';
import { runFixedLayoutTest } from '../__mocks__/run-fixed-layout-test';
import { fixLayout1 } from '../__mocks__/fixed-layout-specs';

const filterStart = (_scope: FlowNodeScope) => !['start'].includes(_scope.meta?.node?.id || '');

const filterEnd = (_scope: FlowNodeScope) => !['end'].includes(_scope.meta?.node?.id || '');

runFixedLayoutTest('Variable Fix Layout Filter Start End', fixLayout1, {
  transformCovers: (scopes) => scopes.filter(filterEnd),
  transformDeps: (scopes) => scopes.filter(filterStart),
});
