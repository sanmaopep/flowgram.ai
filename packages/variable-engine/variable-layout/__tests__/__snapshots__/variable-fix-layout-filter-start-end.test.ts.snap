// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Variable Fix Layout Filter Start End > test get Covers 1`] = `
Map {
  "start" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4",
    "branch_f2c11d0fb42",
    "branch_2c11d0fb42c",
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
    "exclusiveSplit_a59afaadc9a",
    "branch_59afaadc9ac",
    "branch_9afaadc9acd",
    "deleteRecords_c32807e97c5",
    "branch_dbf2c60aee4",
    "updateRecords_7ed2a172c32",
  ],
  "getRecords_07e97c55832" => [
    "forEach_260a8f85ff2",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4",
    "branch_f2c11d0fb42",
    "branch_2c11d0fb42c",
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
    "exclusiveSplit_a59afaadc9a",
    "branch_59afaadc9ac",
    "branch_9afaadc9acd",
    "deleteRecords_c32807e97c5",
    "branch_dbf2c60aee4",
    "updateRecords_7ed2a172c32",
  ],
  "forEach_260a8f85ff2" => [
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
    "exclusiveSplit_a59afaadc9a",
    "branch_59afaadc9ac",
    "branch_9afaadc9acd",
    "deleteRecords_c32807e97c5",
    "branch_dbf2c60aee4",
    "updateRecords_7ed2a172c32",
  ],
  "createRecord_8f85ff2c11d" => [
    "exclusiveSplit_ff2c11d0fb4",
    "branch_f2c11d0fb42",
    "branch_2c11d0fb42c",
  ],
  "exclusiveSplit_ff2c11d0fb4" => [],
  "branch_f2c11d0fb42" => [
    "branch_2c11d0fb42c",
  ],
  "branch_2c11d0fb42c" => [],
  "exclusiveSplit_88dbf2c60ae" => [],
  "branch_8dbf2c60aee" => [
    "branch_dbf2c60aee4",
    "updateRecords_7ed2a172c32",
  ],
  "exclusiveSplit_a59afaadc9a" => [],
  "branch_59afaadc9ac" => [
    "branch_9afaadc9acd",
    "deleteRecords_c32807e97c5",
  ],
  "branch_9afaadc9acd" => [],
  "deleteRecords_c32807e97c5" => [],
  "branch_dbf2c60aee4" => [],
  "updateRecords_7ed2a172c32" => [],
  "end" => [],
}
`;

exports[`Variable Fix Layout Filter Start End > test get Covers After Init Private 1`] = `
Map {
  "start" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4",
    "branch_f2c11d0fb42",
    "branch_2c11d0fb42c",
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
    "exclusiveSplit_a59afaadc9a",
    "branch_59afaadc9ac",
    "branch_9afaadc9acd",
    "deleteRecords_c32807e97c5",
    "branch_dbf2c60aee4",
    "updateRecords_7ed2a172c32",
  ],
  "getRecords_07e97c55832" => [
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
    "createRecord_8f85ff2c11d",
    "createRecord_8f85ff2c11d_private",
    "exclusiveSplit_ff2c11d0fb4",
    "exclusiveSplit_ff2c11d0fb4_private",
    "branch_f2c11d0fb42",
    "branch_f2c11d0fb42_private",
    "branch_2c11d0fb42c",
    "branch_2c11d0fb42c_private",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac",
    "branch_59afaadc9ac_private",
    "branch_9afaadc9acd",
    "branch_9afaadc9acd_private",
    "deleteRecords_c32807e97c5",
    "deleteRecords_c32807e97c5_private",
    "branch_dbf2c60aee4",
    "branch_dbf2c60aee4_private",
    "updateRecords_7ed2a172c32",
    "updateRecords_7ed2a172c32_private",
  ],
  "forEach_260a8f85ff2" => [
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac",
    "branch_59afaadc9ac_private",
    "branch_9afaadc9acd",
    "branch_9afaadc9acd_private",
    "deleteRecords_c32807e97c5",
    "deleteRecords_c32807e97c5_private",
    "branch_dbf2c60aee4",
    "branch_dbf2c60aee4_private",
    "updateRecords_7ed2a172c32",
    "updateRecords_7ed2a172c32_private",
  ],
  "createRecord_8f85ff2c11d" => [
    "exclusiveSplit_ff2c11d0fb4",
    "exclusiveSplit_ff2c11d0fb4_private",
    "branch_f2c11d0fb42",
    "branch_f2c11d0fb42_private",
    "branch_2c11d0fb42c",
    "branch_2c11d0fb42c_private",
  ],
  "exclusiveSplit_ff2c11d0fb4" => [],
  "branch_f2c11d0fb42" => [
    "branch_2c11d0fb42c",
    "branch_2c11d0fb42c_private",
  ],
  "branch_2c11d0fb42c" => [],
  "exclusiveSplit_88dbf2c60ae" => [],
  "branch_8dbf2c60aee" => [
    "branch_dbf2c60aee4",
    "branch_dbf2c60aee4_private",
    "updateRecords_7ed2a172c32",
    "updateRecords_7ed2a172c32_private",
  ],
  "exclusiveSplit_a59afaadc9a" => [],
  "branch_59afaadc9ac" => [
    "branch_9afaadc9acd",
    "branch_9afaadc9acd_private",
    "deleteRecords_c32807e97c5",
    "deleteRecords_c32807e97c5_private",
  ],
  "branch_9afaadc9acd" => [],
  "deleteRecords_c32807e97c5" => [],
  "branch_dbf2c60aee4" => [],
  "updateRecords_7ed2a172c32" => [],
  "end" => [],
}
`;

exports[`Variable Fix Layout Filter Start End > test get Deps 1`] = `
Map {
  "start" => [],
  "getRecords_07e97c55832" => [],
  "forEach_260a8f85ff2" => [
    "getRecords_07e97c55832",
  ],
  "createRecord_8f85ff2c11d" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
  ],
  "exclusiveSplit_ff2c11d0fb4" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "createRecord_8f85ff2c11d",
  ],
  "branch_f2c11d0fb42" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4",
  ],
  "branch_2c11d0fb42c" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4",
    "branch_f2c11d0fb42",
  ],
  "exclusiveSplit_88dbf2c60ae" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
  ],
  "branch_8dbf2c60aee" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
  ],
  "exclusiveSplit_a59afaadc9a" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
  ],
  "branch_59afaadc9ac" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
    "exclusiveSplit_a59afaadc9a",
  ],
  "branch_9afaadc9acd" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
    "exclusiveSplit_a59afaadc9a",
    "branch_59afaadc9ac",
  ],
  "deleteRecords_c32807e97c5" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
    "exclusiveSplit_a59afaadc9a",
    "branch_59afaadc9ac",
    "branch_9afaadc9acd",
  ],
  "branch_dbf2c60aee4" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
  ],
  "updateRecords_7ed2a172c32" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "branch_8dbf2c60aee",
    "branch_dbf2c60aee4",
  ],
  "end" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
  ],
}
`;

exports[`Variable Fix Layout Filter Start End > test get Deps After Init Private 1`] = `
Map {
  "start" => [],
  "getRecords_07e97c55832" => [
    "getRecords_07e97c55832_private",
  ],
  "forEach_260a8f85ff2" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2_private",
  ],
  "createRecord_8f85ff2c11d" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
    "createRecord_8f85ff2c11d_private",
  ],
  "exclusiveSplit_ff2c11d0fb4" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4_private",
  ],
  "branch_f2c11d0fb42" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4",
    "exclusiveSplit_ff2c11d0fb4_private",
    "branch_f2c11d0fb42_private",
  ],
  "branch_2c11d0fb42c" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4",
    "exclusiveSplit_ff2c11d0fb4_private",
    "branch_f2c11d0fb42",
    "branch_2c11d0fb42c_private",
  ],
  "exclusiveSplit_88dbf2c60ae" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae_private",
  ],
  "branch_8dbf2c60aee" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee_private",
  ],
  "exclusiveSplit_a59afaadc9a" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a_private",
  ],
  "branch_59afaadc9ac" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac_private",
  ],
  "branch_9afaadc9acd" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac",
    "branch_9afaadc9acd_private",
  ],
  "deleteRecords_c32807e97c5" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac",
    "branch_9afaadc9acd",
    "branch_9afaadc9acd_private",
    "deleteRecords_c32807e97c5_private",
  ],
  "branch_dbf2c60aee4" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_dbf2c60aee4_private",
  ],
  "updateRecords_7ed2a172c32" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_dbf2c60aee4",
    "branch_dbf2c60aee4_private",
    "updateRecords_7ed2a172c32_private",
  ],
  "end" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
  ],
}
`;

exports[`Variable Fix Layout Filter Start End > test get private scope Covers 1`] = `
Map {
  "start_private" => [
    "start",
  ],
  "getRecords_07e97c55832_private" => [
    "getRecords_07e97c55832",
  ],
  "forEach_260a8f85ff2_private" => [
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
    "createRecord_8f85ff2c11d",
    "createRecord_8f85ff2c11d_private",
    "exclusiveSplit_ff2c11d0fb4",
    "exclusiveSplit_ff2c11d0fb4_private",
    "branch_f2c11d0fb42",
    "branch_f2c11d0fb42_private",
    "branch_2c11d0fb42c",
    "branch_2c11d0fb42c_private",
  ],
  "createRecord_8f85ff2c11d_private" => [
    "createRecord_8f85ff2c11d",
  ],
  "exclusiveSplit_ff2c11d0fb4_private" => [
    "exclusiveSplit_ff2c11d0fb4",
    "exclusiveSplit_ff2c11d0fb4_private",
    "branch_f2c11d0fb42",
    "branch_f2c11d0fb42_private",
    "branch_2c11d0fb42c",
    "branch_2c11d0fb42c_private",
  ],
  "branch_f2c11d0fb42_private" => [
    "branch_f2c11d0fb42",
  ],
  "branch_2c11d0fb42c_private" => [
    "branch_2c11d0fb42c",
  ],
  "exclusiveSplit_88dbf2c60ae_private" => [
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac",
    "branch_59afaadc9ac_private",
    "branch_9afaadc9acd",
    "branch_9afaadc9acd_private",
    "deleteRecords_c32807e97c5",
    "deleteRecords_c32807e97c5_private",
    "branch_dbf2c60aee4",
    "branch_dbf2c60aee4_private",
    "updateRecords_7ed2a172c32",
    "updateRecords_7ed2a172c32_private",
  ],
  "branch_8dbf2c60aee_private" => [
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac",
    "branch_59afaadc9ac_private",
    "branch_9afaadc9acd",
    "branch_9afaadc9acd_private",
    "deleteRecords_c32807e97c5",
    "deleteRecords_c32807e97c5_private",
  ],
  "exclusiveSplit_a59afaadc9a_private" => [
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac",
    "branch_59afaadc9ac_private",
    "branch_9afaadc9acd",
    "branch_9afaadc9acd_private",
    "deleteRecords_c32807e97c5",
    "deleteRecords_c32807e97c5_private",
  ],
  "branch_59afaadc9ac_private" => [
    "branch_59afaadc9ac",
  ],
  "branch_9afaadc9acd_private" => [
    "branch_9afaadc9acd",
    "branch_9afaadc9acd_private",
    "deleteRecords_c32807e97c5",
    "deleteRecords_c32807e97c5_private",
  ],
  "deleteRecords_c32807e97c5_private" => [
    "deleteRecords_c32807e97c5",
  ],
  "branch_dbf2c60aee4_private" => [
    "branch_dbf2c60aee4",
    "branch_dbf2c60aee4_private",
    "updateRecords_7ed2a172c32",
    "updateRecords_7ed2a172c32_private",
  ],
  "updateRecords_7ed2a172c32_private" => [
    "updateRecords_7ed2a172c32",
  ],
  "end_private" => [],
}
`;

exports[`Variable Fix Layout Filter Start End > test get private scope Deps 1`] = `
Map {
  "start_private" => [],
  "getRecords_07e97c55832_private" => [],
  "forEach_260a8f85ff2_private" => [
    "getRecords_07e97c55832",
  ],
  "createRecord_8f85ff2c11d_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
  ],
  "exclusiveSplit_ff2c11d0fb4_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
    "createRecord_8f85ff2c11d",
  ],
  "branch_f2c11d0fb42_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4",
    "exclusiveSplit_ff2c11d0fb4_private",
  ],
  "branch_2c11d0fb42c_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "forEach_260a8f85ff2_private",
    "createRecord_8f85ff2c11d",
    "exclusiveSplit_ff2c11d0fb4",
    "exclusiveSplit_ff2c11d0fb4_private",
    "branch_f2c11d0fb42",
  ],
  "exclusiveSplit_88dbf2c60ae_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
  ],
  "branch_8dbf2c60aee_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
  ],
  "exclusiveSplit_a59afaadc9a_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
  ],
  "branch_59afaadc9ac_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
  ],
  "branch_9afaadc9acd_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac",
  ],
  "deleteRecords_c32807e97c5_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_8dbf2c60aee_private",
    "exclusiveSplit_a59afaadc9a",
    "exclusiveSplit_a59afaadc9a_private",
    "branch_59afaadc9ac",
    "branch_9afaadc9acd",
    "branch_9afaadc9acd_private",
  ],
  "branch_dbf2c60aee4_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
  ],
  "updateRecords_7ed2a172c32_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
    "exclusiveSplit_88dbf2c60ae_private",
    "branch_8dbf2c60aee",
    "branch_dbf2c60aee4",
    "branch_dbf2c60aee4_private",
  ],
  "end_private" => [
    "getRecords_07e97c55832",
    "forEach_260a8f85ff2",
    "exclusiveSplit_88dbf2c60ae",
  ],
}
`;

exports[`Variable Fix Layout Filter Start End > test sort 1`] = `
[
  "start",
  "getRecords_07e97c55832",
  "getRecords_07e97c55832_private",
  "forEach_260a8f85ff2",
  "forEach_260a8f85ff2_private",
  "createRecord_8f85ff2c11d",
  "createRecord_8f85ff2c11d_private",
  "exclusiveSplit_ff2c11d0fb4",
  "exclusiveSplit_ff2c11d0fb4_private",
  "branch_f2c11d0fb42",
  "branch_f2c11d0fb42_private",
  "branch_2c11d0fb42c",
  "branch_2c11d0fb42c_private",
  "exclusiveSplit_88dbf2c60ae",
  "exclusiveSplit_88dbf2c60ae_private",
  "branch_8dbf2c60aee",
  "branch_8dbf2c60aee_private",
  "exclusiveSplit_a59afaadc9a",
  "exclusiveSplit_a59afaadc9a_private",
  "branch_59afaadc9ac",
  "branch_59afaadc9ac_private",
  "branch_9afaadc9acd",
  "branch_9afaadc9acd_private",
  "deleteRecords_c32807e97c5",
  "deleteRecords_c32807e97c5_private",
  "branch_dbf2c60aee4",
  "branch_dbf2c60aee4_private",
  "updateRecords_7ed2a172c32",
  "updateRecords_7ed2a172c32_private",
  "testScope",
  "$blockIcon$exclusiveSplit_ff2c11d0fb4",
  "$inlineBlocks$exclusiveSplit_ff2c11d0fb4",
  "$blockOrderIcon$branch_f2c11d0fb42",
  "$blockOrderIcon$branch_2c11d0fb42c",
  "$blockIcon$exclusiveSplit_88dbf2c60ae",
  "$inlineBlocks$exclusiveSplit_88dbf2c60ae",
  "$blockOrderIcon$branch_8dbf2c60aee",
  "$blockIcon$exclusiveSplit_a59afaadc9a",
  "$inlineBlocks$exclusiveSplit_a59afaadc9a",
  "$blockOrderIcon$branch_59afaadc9ac",
  "$blockOrderIcon$branch_9afaadc9acd",
  "$blockOrderIcon$branch_dbf2c60aee4",
  "end",
  "start_private",
  "end_private",
]
`;
